src/ast/diagnostic.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <cassert>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "location.hh"
+
+namespace bpftrace {
+namespace ast {
+
+// Diagnostic reflects a single error at a single source location. This is a
+// simple wrapper around a string for that message, and the location class.
+class Diagnostic {
+public:
+  enum class Severity {
+    Warning,
+    Error,
+  };
+  Diagnostic(location loc) : loc_(loc) {};
+  const std::string msg() const
+  {
+    return msg_.str();
+  }
+  const std::string hint() const
+  {
+    return hint_.str();
+  }
+  const location& loc() const
+  {
+    return loc_;
+  }
+
+  // Each diagnostic can potentially have a hint attached, which is how to
+  // effectively resolve this issue.
+  std::stringstream& addHint()
+  {
+    return hint_;
+  }
+
+  template <typename T>
+  Diagnostic& operator<<(const T& t)
+  {
+    msg_ << t;
+    return *this;
+  }
+
+private:
+  std::stringstream msg_;
+  std::stringstream hint_;
+  location loc_;
+};
+
+// Diagnostics contains a collection of diagnostics.
+class Diagnostics {
+public:
+  using Severity = Diagnostic::Severity;
+
+  template <typename... Args>
+  Diagnostic& add(Severity severity, Args... args)
+  {
+    auto index = static_cast<long unsigned int>(severity);
+    if (diagnostics_.size() <= index) {
+      diagnostics_.resize(index + 1);
+    }
```

Comment:
Since we have only 2 severity levels, wouldn't it make more sense to resize `diagnostics_` upon creation instead of doing it lazily? I think it would allow to simplify this and some other methods below.

Reply:
Sure, no problem, makes sense.